cid-decode/src/main.rs

```diff
@@ -53,51 +62,136 @@ fn main() -> Result<()> {
         let entry = results.entry(res.clone()).or_default();
         *entry += 1;
 
-        s.clear();
+        buffer.clear();
     }
     // Restore panic hook.
-    panic::set_hook(h);
+    panic::set_hook(default_panic_hook);
 
     // We only get here if reading from Stdin fails...
-    rdr.read_line(&mut s)?;
+    rdr.read_line(&mut buffer)?;
 
     Ok(())
 }
 
 #[derive(Debug, Clone)]
 struct Metadata {
-    base: multibase::Base,
+    base: cid::multibase::Base,
     version: cid::Version,
-    codec: cid::Codec,
-    hash: Option<multihash::Code>,
+    codec: self::codec::Codec,
+    hash: Option<cid::multihash::Code>,
     hash_len: usize,
 }
 
-fn do_single(s: &str) -> Result<Metadata> {
-    let c = cid::Cid::try_from(s)?;
-    if c.version() == cid::Version::V0 {
-        return Ok(Metadata {
-            base: multibase::Base::Base58Btc,
-            version: c.version(),
-            codec: c.codec(),
-            hash: match std::panic::catch_unwind(|| c.hash().algorithm()) {
-                Ok(h) => Some(h),
-                Err(_) => None,
-            },
-            hash_len: c.hash().digest().len(),
-        });
-    }
-
-    let (b, _) = multibase::decode(s)?;
-
-    Ok(Metadata {
-        base: b,
+fn do_single(line: &str) -> Result<Metadata> {
+    let c = cid::Cid::try_from(line)?;
+    return Ok(Metadata {
+        base: if c.version() == cid::Version::V0 {
+            cid::multibase::Base::Base58Btc
+        } else {
+            cid::multibase::decode(line)?.0
```

Comment:
This could use some context, i.e. `decode(...).context("unable to decode multibase")?`. That `context` is in some `failure::ResultExt` I believe, which is probably already imported.

Reply:
Same for the ones below.

Reply:
If I understand the failure crate api correct we are currently not able to use `context()` here. Since we are using `std::result::Result<T, failure:Error>` rather then `failure::ResultExt`. 
I annotated the code with a TODO and suggest that we come back to this when we take on  #1

Reply:
You are right. 
I put some context to the errors.